content/lectures/lec01.md

```diff
@@ -7,91 +7,193 @@ slides: https://docs.google.com/presentation/d/1FVklEogqEGn6zsp8YOpCuynUCCvhB6mX
+  - 172.16.0.0 - 172.31.255.255 (172.16.0.0/12)
+  - 192.168.0.0 - 192.168.255.255 (192.168.0.0/16)
+  
+If you'd like to see your IP address on your machine, you can run one of these commands depending on your OS:
+
+- Linux: `ip addr show`
+- Mac: `ifconfig`
+- Windows: `ipconfig \all`
+
+
+## Visualizing the Network
+
+Say we have two machines on a subnet, this means that these machines know each other exist, and they can communicate, but they don't know _how_ to communicate. They need some shared language that will allow them to send information back and forth between them.
+
+## A Note on IPv6
+
+The four octets in IPv4 only gives us 32 bits of data to specify an IP address, this gives us 2^(32) = 4,294,967,296 possible IP addresses. When IPv4 was invented that seemed like more than we would ever need, but nowadays every smartphone, TV, PlayStation, or laptop might need an IP address. The solution was IPv6 which allowed for 128 bits of data in IP addresses, drastically expanding the number of available IP addresses.
+
+However, IPv6 has really struggled with adoption because network engineers aren't too keen on migrating all of their technology to IPv6. 45% of network traffic in the US happens over IPv6, but that's because many big tech companies have adopted IPv6 on their internal networks. Outside of that, few people have switched over, so we'll use IPv4 throughout this class.
+
+
+## Transport Layer: TCP
+
+TCP is this shared language that is layered on top of IP to allow for a shared language between computers. TCP has robust error-checking to ensure that all of the information transmitted over the internet actually gets to the destination. This error-checking can also cause a delay though as it requires the person receiving the data to also send back a confirmation that they got everything.
+
+TCP also introduces the concept of ports. This is a number ranging from 1-65535 that allows for differentiating between connection types. Typically ports are designated for a specific kind of connection: HTTP is port 80, SSH is port 22, MySQL is port 3306. You can actually configure your machine to listen for whatever kind of connection on whichever port, but you'll confused everyone else who is expecting specific connection types on specific ports.
+
+TCP also allows for multiple concurrent connections. For example, you could have a number of users all connecting to port 80 on an HTTP server and TCP would still be able to manage these connections. This is an essential feature, if we could only establish a single connection the client-server model could never work in networking.
+
+There are some alternatives to TCP:
+- UDP: This protocol is quite similar to TCP but without the error checking. This is great when you want data to be sent fast, and don't care too much about correctness. If you're streaming a show on Netflix, you won't notice if some pixels are flipped during a single frame of the show, so better to be fast than correct.
```

Comment:
This description is correct but the example is wrong. Netflix streaming is over HTTP, so it uses TCP. A better example is VoIP (voice over IP) where you don't care if a frame is garbled but want to make sure frames get to the end user as quickly as possible.

The key distinction here is that error checking would be useless for VoIP because if a packet has to be resent and gets to the end user late, it's useless because you don't want to play someone's sentences out of order.

When people talk about using UDP for streaming, they're frequently talking about live broadcasting with extremely tight requirements around stream delay; not streaming for pre-recorded things.